fava/api/budgets.py

```diff
@@ -0,0 +1,120 @@
+from datetime import datetime
+from dateutil.relativedelta import relativedelta
+from decimal import Decimal
+
+from beancount.core.data import Custom
+
+
+class Budgets(object):
+
+    def __init__(self, entries):
+        """
+        Reads budget directives from the specified list of beancount entries.
+
+        Example for a budget directive:
+
+            2015-04-09 custom "budget" Expenses:Books "monthly"  20.00 EUR
+        """
+        self.datelines = {}
+        self.currencies = set()
+
+        for entry in entries:
+            if isinstance(entry, Custom) and entry.type == 'budget':
+                account_name = entry.values[0].value
+
+                if account_name not in self.datelines:
+                    self.datelines[account_name] = []
+
+                self.datelines[account_name].append(
+                    Dateline(date_start=entry.date,
+                             period=entry.values[1].value,
+                             number=entry.values[2].value.number,
+                             currency=entry.values[2].value.currency)
+                )
+
+                self.currencies.add(entry.values[2].value.currency)
+
+        for name, account in self.datelines.items():
+            self.datelines[name] = sorted(self.datelines[name],
+                                          key=lambda
+                                          dateline: dateline.date_start)
+
+    def _daterange(self, start_date, end_date):
+        """Yields a datetime for every day in the specified interval."""
+        for n in range(int((end_date - start_date).days)):
+            yield start_date + relativedelta(days=n)
+
+    def _matching_dateline(self, datelines, date_active):
+        """Returns the dateline that is active on the specifed date."""
+        last_seen_dateline = None
+        for dateline in datelines:
+            if dateline.date_start <= date_active:
+                last_seen_dateline = dateline
+            else:
+                break
+        return last_seen_dateline
+
+    def budget(self, account_name, date_from, date_to):
+        """
+        Returns a dictionary (currency => number) with the budget for the
+        specified account and period.
+        """
+        currency_dict = {}
+
+        if account_name not in self.datelines.keys():
+            return currency_dict
+
+        for single_day in self._daterange(date_from, date_to):
+            dateline = self._matching_dateline(
+                self.datelines[account_name], single_day)
+
+            if dateline:
+                if dateline.currency not in currency_dict:
+                    currency_dict[dateline.currency] = Decimal(0.0)
+                currency_dict[dateline.currency] += dateline.value(single_day)
+
+        return currency_dict
+
+
+class Dateline(object):
```

Comment:
That this is an object kind of confused me, since half of its functionality seems to be implemented in `Budgets`. I think it would be clearer to have all the date-related only functions in `util.date`, then we would only have `value` left as a method (and it's so trivial, we could just do that computation in the one place where we use it). So maybe just have `Dateline` namedtuple.